alfalfa_worker/lib/models.py

```diff
@@ -0,0 +1,262 @@
+    FloatField,
+    IntField,
+    ListField,
+    ReferenceField,
+    StringField,
+    UUIDField
+)
+
+from alfalfa_worker.lib.sim_type import SimType
+
+
+class TimestampedDocument(Document):
+    """Create a class to save the created_at and updated_at fields automatically"""
+    meta = {'allow_inheritance': True, 'abstract': True}
+
+    created = DateTimeField(required=True, default=datetime.datetime.now)
+    modified = DateTimeField(required=True, default=datetime.datetime.now)
+
+    def save(self, force_insert=False, validate=True, clean=True, write_concern=None, cascade=None,
+             cascade_kwargs=None, _refs=None, save_condition=None, signal_kwargs=None, **kwargs,):
+        self.modified = datetime.datetime.now()
+        return super().save(
+            force_insert, validate, clean, write_concern, cascade, cascade_kwargs, _refs, save_condition,
+            signal_kwargs, **kwargs
+        )
+
+    def to_dict(self):
+        """Override the to_dict method to include to convert specific fields correctly.
+        e.g., created and modified should be converted to iso8601 strings.
+        """
+        created = self.created.isoformat()
+        modified = self.modified.isoformat()
+
+        data = json.loads(self.to_json())
+
+        data['created'] = created
+        data['modified'] = modified
+        return data
+
+
+class Site(TimestampedDocument):
+    meta = {'collection': 'site'}
+
+    # external reference ID
+    ref_id = StringField(required=True, index=True)
+    name = StringField(required=True, max_length=255)
+    haystack_raw = ListField(DictField())
+
+    # In old database the first recs is the site, this is now moved to the
+    # site object in the database. Need to type this eventually
+    dis = StringField()
+    site = StringField()
+    area = StringField()
+    geo_city = StringField()
+    geo_coord = StringField()
+    geo_country = StringField()
+    geo_state = StringField()
+    tz = StringField()
+    weather_ref = StringField()
+    sim_status = StringField()
+    sim_type = StringField()
+
+    def recs(self, **kwargs: dict):  # -> list[Rec]:
+        """Return list of all the recs for the site
+
+        Args:
+            **kwargs: Additional arguments to pass to the Rec.objects.filter() method.
+
+        Returns:
+            list: Rec objects
+
+        """
+        return Rec.objects.filter(site=self, **kwargs)
+
+
+class RecInstance(EmbeddedDocument):
+    """This is a flat haystack representation of a point. Right now this stores the union of any
+    value that is expected. Many of these fields are probably not used anymore on this object
+    (e.g., simStatus, step), but they are still included because the code base is reading/writing
+    the valued. For example, it is unsetting `step` even though it doesn't appear to be written to."""
+
+    # TODO: maybe we need to make this a dynamic field: https://docs.mongoengine.org/guide/defining-documents.html#dynamic-document-schemas
```

Comment:
this is an important note. If there is a haystack field that isn't here, then alfalfa won't allow it to persist.

Reply:
I wonder if instead of trying to capture an entire all of the possible tags we instead have a `Tag` model which each just have a key, value, and a reference to a `Rec`. I feel like this could also potentially make the transition to postgres easier since it's a common relational database paradigm.